### PR TITLE
Support and embed payload UUIDs

### PR DESCRIPTION
**Pull request by hdm**
This changeset adds support for embedding UUIDs into payloads and displaying the connected UUID in the `sessions -l -v` output. This PR handles most of the stagers and payloads that work over HTTP & HTTPS. This work is backwards-compatible with older stagers (64-bit reverse_http for example).

In addition, this PR adds two new datastore options for stagers and payloads using the HTTP/HTTPS transports:
- `PayloadUUIDSeed`: A human-readable string that is hashed and used to calculate a Payload UID
- `PayloadUUIDRaw`: An 8-byte raw Payload UID specified as a 16-byte hex string

A UUID is made up of 4 distinct fields:
- `puid`: The Payload UID is a possibly-random 8-byte prefix identifying a specific generated payload. This is randomly generated or derived from `PayloadUUIDSeed` or `PayloadUUIDRaw`
- `arch`: The Architecture ID corresponds to a table of constants that map to every architecture known within Metasploit. This isn't perfect due to weirdness within Metasploit (X64 isn't exactly X86_64)
- `platform`: The Platform ID corresponds to a table of constants that map to every platformknown within Metasploit. If the payload supports multiple platforms, the first one in the PlatformList will be stored in the UUID.
- `timestamp`: The Timestamp is a Unix time value (UTC) of when the payload was generated. Values prior to 2015-01-01 or too far into the future are treated as invalid and ignored.

These four values are then obfuscated using two random bytes via XOR. These two random bytes result in a fairly random 16-byte UUID value. For HTTP/HTTPS transports, these 16 bytes are further encoded as a 22-byte Base64URL string. 

Generated stagers and payloads now automatically embed a UUID if there is enough room. In the case of HTTP/HTTPS transports, the first 22 bytes of the URL (excluding non-Base64URL bytes) are treated as the UUID. If the UUID looks "invalid" (bad arch, platform, timestamp), only the Payload UID will be extracted, preserving backwards compatibility with existing payloads.

Stagers using HTTP/HTTPS transports will reuse the UUID during the switch from "init" to "connection" URLs, preserving the UUID between Metasploit instances and the staging process. The currently associated UUID can be seed with the `-v` flag to the `sessions -l` command.

There is still quite a bit of work left to do before UUIDs become practical, but this PR creates the initial path, and opens the door to smaller future PRs to add support to specific stagers and payloads. The end goal would be to enable universal handlers that can detect the architecture and platform on the fly and make it easy to track exactly what payload connected back to a shared listener.

### CODE REVIEW

File: lib/rex/payloads/meterpreter/uri_checksum.rb [hunk 1]
```diff
@@ -9,119 +11,123 @@ module UriChecksum
...
-          case uri_check
-            when URI_CHECKSUM_INITW
-              uri_match = "/INITM"
-            when URI_CHECKSUM_INITP
-              uri_match = "/INITPY"
-            when URI_CHECKSUM_INITJ
-              uri_match = "/INITJM"
-            when URI_CHECKSUM_CONN
-              uri_match = "/CONN_" + ( uri_conn || Rex::Text.rand_text_alphanumeric(16) )
+        # @param uri_match [String] The URI string from the HTTP request
```
Comment (void-in) [inline lib/rex/payloads/meterpreter/uri_checksum.rb]: Should be
```ruby
@param uri [String] The URI string from the HTTP request
```
Currently the parameter is specified as uri_match

**Comment by hmoore-r7** [comment]
Testing is fairly straight-forward. Generate a new stager or payload. Run a handler. Execute the payload. Look at ```sessions -l -v```. Close the session, run this again, and confirm that the UUID changes. For Meterpreter, the UUID should be preserved after a transport change and a migration. This change shouldn't break any existing functionality. Setting a PayloadUUIDSeed or PayloadUUIDRaw value and trying to use a non-dynamic stager (64-bit Windows reverse_https) should result in an argument error at runtime.

**Comment by OJ** [comment]
*_rolls up sleeves_* here we go!

File: lib/msf/base/serializer/readable_text.rb [hunk 2]
```diff
@@ -555,7 +556,11 @@ def self.dump_sessions(framework, opts={})
       if session.respond_to? :platform
         row[1] += " " + session.platform
       end
-      row << session.via_exploit if verbose and session.via_exploit
+
+      if verbose
+        row << session.via_exploit.to_s
+        row << session.payload_uuid.to_s
```
Comment (OJ) [inline lib/msf/base/serializer/readable_text.rb]: Will `via_exploit` or `payload_uuid` ever be `nil`? Should we do this instead, to be safe:
``` ruby
row << (session.via_exploit || '').to_s
row << (session.payload_uuid || '').to_s
```
I'm guessing `payload_uuid` will be fine, given the changes we're making here. But the history of `via_exploit` had a check for `nil` (even if it was incorrect!). Should we be maintaining it as above?

File: lib/msf/core/handler/reverse_http.rb [hunk 3]
```diff
@@ -215,17 +215,25 @@ def lookup_proxy_settings
   #
   def on_request(cli, req, obj)
     resp = Rex::Proto::Http::Response.new
+    info = process_uri_resource(req.relative_resource)
+    uuid = info[:uuid] || Msf::Payload::UUID.new
 
-    print_status("#{cli.peerhost}:#{cli.peerport} Request received for #{req.relative_resource}...")
+    # Configure the UUID architecture and payload if necessary
+    uuid.arch      = obj.arch if uuid.arch.nil?
+    uuid.platform  = obj.platform if uuid.platform.nil?
```
Comment (OJ) [inline lib/msf/core/handler/reverse_http.rb]: Is the usual approach to use `||=`?
``` ruby
uuid.arch ||= obj.arch
uuid.platform ||= obj.platform
```

Seems to be quite common:
```
$ grep -r "||=" . | wc -l
1824
```
To be clear, I don't really mind :)

File: lib/msf/core/handler/reverse_http.rb [hunk 4]
```diff
@@ -215,17 +215,25 @@ def lookup_proxy_settings
...
+    end
 
     self.pending_connections += 1
 
     # Process the requested resource.
-    case uri_match
-      when /^\/INITPY/
-        conn_id = generate_uri_checksum(URI_CHECKSUM_CONN) + "_" + Rex::Text.rand_text_alphanumeric(16)
+    case info[:mode]
+      when :init_python
```
Comment (OJ) [inline lib/msf/core/handler/reverse_http.rb]: Love the use of symbols here instead of URIs.

File: lib/rex/post/meterpreter/packet_dispatcher.rb [hunk 5]
```diff
@@ -66,9 +66,12 @@ def initialize_passive_dispatcher
     self.waiters    = []
     self.alive      = true
 
+    # Ensure that there is only one leading and trailing slash on the URI
+    resource_uri = "/" + self.conn_id.to_s.gsub(/(^\/|\/$)/, '') + "/"
```
Comment (OJ) [inline lib/rex/post/meterpreter/packet_dispatcher.rb]: Given that it's possible for Base64 encoded stuff to include slashes, and they _could_ appear at the end, should we do something to protect against this? eg.:
``` ruby
2.1.5 :002 > puts Base64.encode64("subjects?")
c3ViamVjdHM/
```
I realise it's a pretty slim chance, but should we care?

Comment (OJ) [inline lib/rex/post/meterpreter/packet_dispatcher.rb]: Ah, I see below there's substitution going. Nevermind!

**Comment by OJ** [comment]
Hard to find anything bad in this as far as the code goes. Will get stuck into testing pronto.

**Comment by OJ** [comment]
```
msf exploit(handler) > sessions -v

Active sessions
===============

  Id  Type                       Information                                  Connection                                        Via                    PayloadId
  --  ----                       -----------                                  ----------                                        ---                    ---------
  1   meterpreter x86/win32      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8443 -> 10.1.10.45:61644 (10.1.10.45)  exploit/multi/handler  f15e59ce95a18f81/x86=1/windows=1/2015-04-06T06:42:10Z
  2   meterpreter x86/win32      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8443 -> 10.1.10.45:61646 (10.1.10.45)  exploit/multi/handler  6933f4732929cd80/x86=1/windows=1/2015-04-06T06:43:15Z
  3   meterpreter x86/win32      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8443 -> 10.1.10.45:61648 (10.1.10.45)  exploit/multi/handler  207d272536682bd6/x86=1/windows=1/2015-04-06T06:43:23Z
  4   meterpreter x64/win64      WIN-S45GUQ5KGVK\OJ @ WIN-S45GUQ5KGVK         10.1.10.40:8000 -> 10.1.10.35:62379 (10.1.10.35)  exploit/multi/handler  54726fba951eb83c/x86_64=2/windows=1/2015-04-06T06:47:24Z
  5   meterpreter x64/win64      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8000 -> 10.1.10.45:61661 (10.1.10.45)  exploit/multi/handler  563bdf606c36341a/x86_64=2/windows=1/2015-04-06T06:48:16Z
  6   meterpreter python/python  oj @ ropchain                                10.1.10.40:5000 -> 10.1.10.40:40940 (10.1.10.40)  exploit/multi/handler  032f24e529ca13b0/python=20/python=21/2015-04-06T06:49:51Z
  7   meterpreter python/python  oj @ ropchain                                10.1.10.40:5000 -> 10.1.10.40:41151 (10.1.10.40)  exploit/multi/handler  032f24e529ca13b0/python=20/python=21/2015-04-06T06:51:27Z
  8   meterpreter python/python  WIN-S45GUQ5KGVK\OJ @ WIN-S45GUQ5KGVK         10.1.10.40:5000 -> 10.1.10.35:62383 (10.1.10.35)  exploit/multi/handler  032f24e529ca13b0/python=20/python=21/2015-04-06T06:53:27Z
```

**Comment by OJ** [comment]
Migrating:
```
msf exploit(handler) > sessions -i 4
[*] Starting interaction with 4...

meterpreter > sysinfo
Computer        : WIN-S45GUQ5KGVK
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64
System Language : en_US
Meterpreter     : x64/win64
meterpreter > migrate 2224
[*] Migrating from 4696 to 2224...
[*] Migration completed successfully.
meterpreter > sysinfo
Computer        : WIN-S45GUQ5KGVK
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64 (Current Process is WOW64)
System Language : en_US
Meterpreter     : x86/win32
meterpreter > background
[*] Backgrounding session 4...
msf exploit(handler) > sessions -v

Active sessions
===============

  Id  Type                       Information                                  Connection                                        Via                    PayloadId
  --  ----                       -----------                                  ----------                                        ---                    ---------
  1   meterpreter x86/win32      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8443 -> 10.1.10.45:61644 (10.1.10.45)  exploit/multi/handler  f15e59ce95a18f81/x86=1/windows=1/2015-04-06T06:42:10Z
  2   meterpreter x86/win32      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8443 -> 10.1.10.45:61646 (10.1.10.45)  exploit/multi/handler  6933f4732929cd80/x86=1/windows=1/2015-04-06T06:43:15Z
  3   meterpreter x86/win32      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8443 -> 10.1.10.45:61648 (10.1.10.45)  exploit/multi/handler  207d272536682bd6/x86=1/windows=1/2015-04-06T06:43:23Z
**  4   meterpreter x86/win32      WIN-S45GUQ5KGVK\OJ @ WIN-S45GUQ5KGVK         10.1.10.40:8000 -> 10.1.10.35:62379 (10.1.10.35)  exploit/multi/handler  54726fba951eb83c/x86_64=2/windows=1/2015-04-06T06:47:24Z
  5   meterpreter x64/win64      WIN-8RDFKU33NLH\OJ Reeves @ WIN-8RDFKU33NLH  10.1.10.40:8000 -> 10.1.10.45:61661 (10.1.10.45)  exploit/multi/handler  563bdf606c36341a/x86_64=2/windows=1/2015-04-06T06:48:16Z
  6   meterpreter python/python  oj @ ropchain                                10.1.10.40:5000 -> 10.1.10.40:40940 (10.1.10.40)  exploit/multi/handler  032f24e529ca13b0/python=20/python=21/2015-04-06T06:49:51Z
  7   meterpreter python/python  oj @ ropchain                                10.1.10.40:5000 -> 10.1.10.40:41151 (10.1.10.40)  exploit/multi/handler  032f24e529ca13b0/python=20/python=21/2015-04-06T06:51:27Z
  8   meterpreter python/python  WIN-S45GUQ5KGVK\OJ @ WIN-S45GUQ5KGVK         10.1.10.40:5000 -> 10.1.10.35:62383 (10.1.10.35)  exploit/multi/handler  032f24e529ca13b0/python=20/python=21/2015-04-06T06:53:27Z
```
\*\* Session is now x86, but the payload ID remains the same as before (as HD documented in the PR).

**Comment by OJ** [comment]
General testing of things seems to be fine. I haven't seen any misbehaviours that I wouldn't expect to see.

**Comment by OJ** [comment]
We do have a little bit of strangeness going on with stageless https:

Set up handler:
```
msf exploit(handler) > run

[*] Started HTTPS reverse handler on https://0.0.0.0:6050/
[*] Starting the payload handler...
```
Fire off a single stagless payload, and this happens:
```
[*] 10.1.10.35:63828 Request received for /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs... (UUID:966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z)
[*] Incoming orphaned or stageless session /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs, attaching...
[*] Meterpreter session 3 opened (10.1.10.40:6050 -> 10.1.10.35:63828) at 2015-04-06 17:28:37 +1000
[*] 10.1.10.35:63829 Request received for /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs... (UUID:966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z)
[*] Incoming orphaned or stageless session /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs, attaching...
[*] Meterpreter session 4 opened (10.1.10.40:6050 -> 10.1.10.35:63829) at 2015-04-06 17:28:37 +1000
[*] 10.1.10.35:63830 Request received for /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs... (UUID:966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z)
[*] Incoming orphaned or stageless session /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs, attaching...
[*] Meterpreter session 5 opened (10.1.10.40:6050 -> 10.1.10.35:63830) at 2015-04-06 17:28:37 +1000
[*] 10.1.10.35:63831 Request received for /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs... (UUID:966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z)
[*] Incoming orphaned or stageless session /lmtKZx0J_E9YMVkwDRNrMAAE2Sja3Ma4s5TkcT5JFNuBTbuei_U_9o2-ZBBNByda9lkRZA8d8kFXs, attaching...

[*] Meterpreter session 6 opened (10.1.10.40:6050 -> 10.1.10.35:63831) at 2015-04-06 17:28:38 +1000
meterpreter > background
```
If the handler has `ExitOnSession` set to `false` the sessions will spawn indefinitely. None of the them are responsive
```
msf exploit(handler) > sessions -v

Active sessions
===============

  Id  Type                   Information  Connection                                        Via                    PayloadId
  --  ----                   -----------  ----------                                        ---                    ---------
  3   meterpreter x86/win32               10.1.10.40:6050 -> 10.1.10.35:63828 (10.1.10.35)  exploit/multi/handler  966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z
  4   meterpreter x86/win32               10.1.10.40:6050 -> 10.1.10.35:63829 (10.1.10.35)  exploit/multi/handler  966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z
  5   meterpreter x86/win32               10.1.10.40:6050 -> 10.1.10.35:63830 (10.1.10.35)  exploit/multi/handler  966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z
  6   meterpreter x86/win32               10.1.10.40:6050 -> 10.1.10.35:63831 (10.1.10.35)  exploit/multi/handler  966b4a671d09fc4f/x86=1/windows=1/2015-04-06T07:17:21Z
```
Investigating.

**Comment by OJ** [comment]
Gotcha. Trailing slash was missing. [PR for you](https://github.com/hmoore-r7/metasploit-framework/pull/11) @hmoore-r7.

File: lib/msf/base/serializer/readable_text.rb (hunk 2 above)
Comment (hmoore-r7) [inline lib/msf/base/serializer/readable_text.rb]: In the nil case, the code above will still work correctly (nil.to_s == "")

File: lib/rex/post/meterpreter/packet_dispatcher.rb (hunk 5 above)
Comment (hmoore-r7) [inline lib/rex/post/meterpreter/packet_dispatcher.rb]: We also use Base64Url (vs Base64), which transcribes slashes and and plus signs to - and _ instead.

File: lib/msf/core/handler/reverse_http.rb (hunk 3 above)
Comment (hmoore-r7) [inline lib/msf/core/handler/reverse_http.rb]: Good point!

**Comment by hmoore-r7** [comment]
Stageless payloads should not be using the "hashedstring_CONNID" syntax, but long random URLs. It looks like more work needs to be done here. Maybe this is the case of an older binary not being handled nicely?

**Comment by OJ** [comment]
That was definitely done with new binaries, as I didn't want to make the same mistake as before!
The URIs do seem like they're slightly different to the URIs generated in staged connections. I'll do a bit more validation around this.

**Comment by hmoore-r7** [comment]
re: Different URIs. The old-style connect URIs were still used prior to this PR. This PR should be setting longer URIs in stageless, but any EXEs before this PR would still have the old format.

**Comment by OJ** [comment]
Looking good this end. Thanks for the updates. Landing now.

**Pull request merged**